hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueService.java

```diff
@@ -390,130 +383,62 @@ public String getQuorumName(String name) {
-        private final Semaphore semaphore = new Semaphore(0);
-        private final ExecutionCallback<Object> mergeCallback = new ExecutionCallback<Object>() {
-            @Override
-            public void onResponse(Object response) {
-                semaphore.release(1);
-            }
-
-            @Override
-            public void onFailure(Throwable t) {
-                logger.warning("Error while running queue merge operation: " + t.getMessage());
-                semaphore.release(1);
-            }
-        };
-
-        private final Map<Integer, Map<QueueContainer, List<QueueItem>>> itemMap;
+    private class Merger extends AbstractContainerMerger<QueueContainer> {
 
-        Merger(Map<Integer, Map<QueueContainer, List<QueueItem>>> itemMap) {
-            this.itemMap = itemMap;
+        Merger(QueueContainerCollector collector) {
+            super(collector, nodeEngine);
         }
 
         @Override
-        public void run() {
-            // we cannot merge into a 3.9 cluster, since not all members may understand the QueueMergeOperation
-            // RU_COMPAT_3_9
-            if (nodeEngine.getClusterService().getClusterVersion().isLessThan(Versions.V3_10)) {
-                logger.info("Cluster needs to run version " + Versions.V3_10 + " to merge queue instances");
-                return;
-            }
+        protected String getLabel() {
+            return "queue";
+        }
 
-            int itemCount = 0;
-            int operationCount = 0;
+        @Override
+        public void runInternal() {
             List<MergingValueHolder<Data>> mergingValues;
-            for (Entry<Integer, Map<QueueContainer, List<QueueItem>>> partitionMap : itemMap.entrySet()) {
-                int partitionId = partitionMap.getKey();
-                Map<QueueContainer, List<QueueItem>> containerMap = partitionMap.getValue();
-                for (Entry<QueueContainer, List<QueueItem>> entry : containerMap.entrySet()) {
-                    QueueContainer container = entry.getKey();
-                    List<QueueItem> itemList = entry.getValue();
+            for (Entry<Integer, Collection<QueueContainer>> entry : collector.getCollectedContainers().entrySet()) {
+                int partitionId = entry.getKey();
+                Collection<QueueContainer> containerList = entry.getValue();
+                for (QueueContainer container : containerList) {
+                    Deque<QueueItem> itemList = container.getItemQueue();
 
-                    int batchSize = container.getConfig().getMergePolicyConfig().getBatchSize();
-                    SplitBrainMergePolicy mergePolicy = getMergePolicy(container);
                     String name = container.getName();
+                    int batchSize = container.getConfig().getMergePolicyConfig().getBatchSize();
+                    SplitBrainMergePolicy mergePolicy = getMergePolicy(container.getConfig().getMergePolicyConfig());
```

Comment:
Minor: you can use `((QueueContainerCollector) collector).getMergePolicyConfig(container)` here as well. Same in `CollectionService` and `MultiMapService`, possibly others as well.

Reply:
Those containers know their configuration, that workaround via the collector was just meant for the lean containers of `AtomicLong`, `AtomicReference` and `CardinalityEstimator`, which cannot provide that info on their own.

I also think the code is easier to read without the casting.
```java
mergePolicy = getMergePolicy(container.getConfig().getMergePolicyConfig());
```
vs.
```java
mergePolicy = getMergePolicy(((QueueContainerCollector) collector).getMergePolicyConfig(container));
```
The collector would anyway call `container.getConfig().getMergePolicyConfig()` again. I guess we can save HotSpot a tiny bit of work here.